### PR TITLE
[f] 501 - abstract twitter authentication and authorization

### PR DESCRIPTION
**Pull request by giang271291**
Description
--
* Abstract authentication steps so that it is not coupled with the underlying auth library

How to test
--
* Start twitterbot with incorrect `consumer_key`, `consumer_secret`, `token_key`, `token_secret` configurations (either from file or CLI) should raise an AuthenticationError

### CODE REVIEW

File: tests/unit_tests/utils/test_auth_utils.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+from twitterbot.utils import auth_utils
+
+
+class AuthUtilsTestCase(TestCase):
+    def test_failed_auth(self):
+        with patch('twitterbot.utils.auth_utils.tweepy.API.verify_credentials',
+                   side_effect=TweepError(reason=[{'message': 'some message'}])):
+            self.assertRaises(AuthenticationError, auth_utils.auth,
+                              consumer_key='ck', consumer_secret='cs', token_key='tk', token_secret='ts'
+            )
```
Comment (tklarryonline) [inline tests/unit_tests/utils/test_auth_utils.py]: The `)` seems to be very lonely. If you are using this style, you can do like 

```py
self.assertRaises(
    AuthenticationError,
    auth_utils.auth,
    consumer_key='ck', consumer_secret='cs', token_key='tk', token_secret='ts'
)
```

File: tests/unit_tests/utils/test_auth_utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    @patch('twitterbot.utils.auth_utils.tweepy.API.verify_credentials')
+    @patch('twitterbot.utils.auth_utils.tweepy.API.me', return_value=MagicMock(screen_name='me'))
+    def test_successful_auth(self, mock_me, mock_verify):
+        try:
+            client = auth_utils.auth(consumer_key='ck', consumer_secret='cs', token_key='tk', token_secret='ts')
+
+            self.assertIsInstance(client, TweetClient)
+        except:
+            self.fail('Expected no exception')
```
Comment (tklarryonline) [inline tests/unit_tests/utils/test_auth_utils.py]: Hmm, is it a little bit over careful to catch the Exception and make it a Fail here?

On the other hand, I think the check for `TweetClient` instance should be put after the try...except block

File: twitterbot/twitterbot.py [hunk 3]
```diff
@@ -1,24 +1,35 @@
...
+    Twitter bot implementation
+    """
     def __init__(self, *args, **kwargs):
         self.config = TwitterBotConfig(*args, **kwargs)
 
-    def start(self, *args, **kwargs):
+    def start(self):
+        """
+        Init and start twitter bot
+        :return: nothing
```
Comment (tklarryonline) [inline twitterbot/twitterbot.py]: There is nothing wrong with Sphinx if you delete the `:return:` key in the doc, I think.

Comment (anhhuy1605) [inline twitterbot/twitterbot.py]: Yeah, we will have a look around to see if other open source packages put return nothing in their document.

**Comment by tklarryonline** [comment]
Tested and worked. However, I feel uneasy about this traceback:

```py
Twitterbot started
Traceback (most recent call last):
  File "/path/to/twitterbot/build/lib/twitterbot/utils/auth_utils.py", line 30, in auth
    client.verify_credentials()
  File "/path/to/.virtualenvs/twitterbot/lib/python3.5/site-packages/tweepy/api.py", line 631, in verify_credentials
    )(**kargs)
  File "/path/to/.virtualenvs/twitterbot/lib/python3.5/site-packages/tweepy/binder.py", line 245, in _call
    return method.execute()
  File "/path/to/.virtualenvs/twitterbot/lib/python3.5/site-packages/tweepy/binder.py", line 229, in execute
    raise TweepError(error_msg, resp, api_code=api_error_code)
tweepy.error.TweepError: [{'code': 89, 'message': 'Invalid or expired token.'}]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/path/to/.virtualenvs/twitterbot/bin/start_twitterbot", line 9, in <module>
    load_entry_point('twitterbot', 'console_scripts', 'start_twitterbot')()
  File "/path/to/.virtualenvs/twitterbot/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/path/to/.virtualenvs/twitterbot/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/path/to/.virtualenvs/twitterbot/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/path/to/.virtualenvs/twitterbot/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/path/to/twitterbot/build/lib/twitterbot/start_twitterbot.py", line 19, in main
    TwitterBot(*args, **kwargs).start()
  File "/path/to/twitterbot/build/lib/twitterbot/twitterbot.py", line 30, in start
    token_secret=self.config.get('token_secret'),
  File "/path/to/twitterbot/build/lib/twitterbot/utils/auth_utils.py", line 32, in auth
    raise AuthenticationError('Tweepy error: %s' % e.args[0][0]['message'])
twitterbot.common.exceptions.AuthenticationError: Tweepy error: Invalid or expired token.
```

File: twitterbot/listeners/base.py [hunk 4]
```diff
@@ -1,8 +1,17 @@
...
+    """
     def __init__(self):
         self.handlers = []
 
     def add_handler(self, handler):
+        """
+        Register a discovered handler to this listener
+        :param handler: a handler instance
+        :type handler: a class derived from BaseTweethandler
+        :return: nothing
```
Comment (tklarryonline) [inline twitterbot/listeners/base.py]: This line is not yet removed.

File: twitterbot/twitterbot.py (hunk 3 above)
Comment (tklarryonline) [inline twitterbot/twitterbot.py]: I saw that you have removed some of the `:return: thing` lines in latest commit. But not this one.

**Comment by tklarryonline** [comment]
Looks good to me now.

**Pull request merged**